Allow resizing Branch and Push/Pull toolbar buttons

jpedroso opened the pull request:
Closes #4569
Closes #17388

## Description

Allow the user to resize the Branch and Push/Pull toolbar buttons, and their dropdown panels.

- Both buttons have a minimum width of `230px`, same as the sidebar. The max width of each button is determined by the available space in the toolbar after taking out the sidebar, minus the other button’s width. This is controlled by `IConstrainedValue` values and determined in `updateResizableConstraints()`.

- If the user-defined width of both buttons is wider than the toolbar's available width the Branch toolbar button takes precedence over the Push/Pull button. For example, when the window resizes, the Push/Pull button’s width is reduced before the Branch button’s.

- Like the sidebar, their width can be reset to the original value (`230px`) with a double-click on the resizable edge.

- The dropdown panels of each button follow the width of the button. The Branches/Pull Requests dropdown has a minimum width of 365px—the previous fixed width—to avoid the panel getting too narrow if the branches names are not long enough.

- The empty states for the Branches and Pull Requests tabs are adjusted to the flexible width. They are now centered in the container. Decided to give them a fixed width of `365px`, same as  previous size, to maintain the aspect of their contents. Letting them flex to full width would make the imagery scale too much and the call to action button quite wide.

### Screenshots

#### Walkthrough of resizing the Branch and Push/Pull toolbar buttons, Branches/PR dropdown and empty states
https://github.com/desktop/desktop/assets/3560/e42eec13-2742-4879-b1f5-0c81d5fae1ea

#### Branches dropdown empty state
![2024-01-28-002297@2x](https://github.com/desktop/desktop/assets/3560/81abf8a3-323a-43ac-8ba1-b6fdd5726d9b)

#### Minimum window size and maximum zoom
![2024-01-28-002295@2x](https://github.com/desktop/desktop/assets/3560/93f73118-6e71-4047-81df-859dbd3b6226)

#### Max zoom, branches dropdown empty state
![2024-01-28-002296@2x](https://github.com/desktop/desktop/assets/3560/fa4dd155-548b-4c5c-bfb2-ae9c3b98d8c2)

## Release notes

Allow the user to resize the Branches and Push/Pull toolbar buttons.

jpedroso committed:
Make toolbar branch dropdown button resizable

jpedroso committed:
Avoid shrink on toolbar-dropdown-arrow-button

Prevents the dropdown arrow button from shrinking when the window or toolbar items are resized.

jpedroso committed:
Update the max width available for the branch dropdown

Set a new constrained value for the branch dropdown so it can only be as wide as the available space after taking the sidebar and pull fetch button widths.

The minimum size remains constant at its default width.

jpedroso committed:
Use branch dropdown's constrained value min and max in resizable

Instead of using a hardcoded max width of 600 and the default min width of 200, the branch dropdown resizable now uses the min and max that are provided with the constrained value.

jpedroso committed:
Rename toolbarButtonsWidth to toolbarButtonsMinWidth

This local var needs to reflect that some of the toolbar buttons are now resizable, and what is represents the total min width of those buttons.

jpedroso committed:
Use a constant for the push/fetch button width

Replace the inline hardcoded value with a constant defined closer the other toolbar constants.

jpedroso committed:
Fix typo in documentation

jpedroso committed:
Rename vars related to push/pull button

More consistent with the `PushPullButton` type they relate to.

jpedroso committed:
Make push pull toolbar button resizable

jpedroso committed:
Center empty states in branches and pull requests container

When there are no branches or pull requests to show, the corresponding empty states are centered on the container with a fixed width of 365px.

For the branches empty state container, the fixed width replaces `flex: 1` to prevent it from filling the container and make the Create Branch button too wide.

jpedroso committed:
Set min width for the branches container

Prevents the branches dropdown container from becoming too narrow when branch names are not long enough.

jpedroso committed:
Add vertical spacing around branches and pr list empty states

With the empty states centered on the container, their imagery would get uncomfortably close to the top. This introduces some spacing on top and bottom to make layout more balanced.

jpedroso committed:
Documentation for Branch dropdown dispatchers

jpedroso commented on app/src/ui/toolbar/dropdown.tsx:
```diff
@@ -352,7 +352,7 @@ export class ToolbarDropdown extends React.Component<
     }
 
     return {
-      position: 'absolute',
+      position: 'fixed',
```
Required to correctly place the dropdown below its toolbar button.

jpedroso replied:
Changing the default properties of the foldout style here might not be the best approach. I’m thinking the branch dropdown should instead provide its own `foldoutStyle` similar to what the sidebar dropdown has.

jpedroso commented on app/styles/ui/_branches.scss:
```diff
@@ -4,14 +4,15 @@
   display: flex;
   flex-direction: column;
-  width: 365px;
+  min-width: 365px;
+  width: 100%;
```
Replace the fixed width with minimum width so the branch toolbar dropdown always has at least 365px. If the toolbar button is wider than that, allow it to fill the same width.

jpedroso commented on app/styles/ui/_branches.scss:
```diff
@@ -4,14 +4,15 @@
 
   .branches-list {
-    width: 365px;
+    width: 100%;
```
Replace the fixed width and allow the branch list to fill the dropdown container (`.branches-container`).

jpedroso commented on app/styles/ui/_branches.scss:
```diff
@@ -198,6 +199,7 @@
 
 .no-pull-requests {
   width: 365px;
+  margin: var(--spacing) auto;
```
Center the empty state content of the Pull Requests tab in the dropdown container, and give its imagery some spacing vertically. 

Read more about the why center empty states below in the Branches tab empty state comments.

jpedroso commented on app/styles/ui/_no-branches.scss:
```diff
@@ -1,6 +1,7 @@
 .no-branches {
+  width: 365px;
+  margin: var(--spacing) auto;
   display: flex;
-  flex: 1;
```
Both for the Branches and Pull Requests empty states, I decided to keep the fixed width of 365px instead of letting the imagery, text and button expand to fill the container. Specifically, in the case of Branches tab, the Create New Branch could get uncomfortably wide.  See comparison below.


#### Full width empty state
![2024-01-28-002298@2x](https://github.com/desktop/desktop/assets/3560/f974965a-a6df-4006-a505-b90ced01d09f)

#### Fixed width empty state
![2024-01-28-002299@2x](https://github.com/desktop/desktop/assets/3560/29ad7c55-0912-4e36-b1ed-db22174829da)

jpedroso commented on app/styles/ui/toolbar/_toolbar.scss:
```diff
@@ -33,19 +33,20 @@
     }
     &.toolbar-dropdown-arrow-button {
       width: 39px;
+      flex-shrink: 0;
```
Prevent shrinking the dropdown arrow button. Keep them always pinned to the right.

jpedroso commented on app/styles/ui/toolbar/_toolbar.scss:
```diff
@@ -33,19 +33,20 @@
     &.push-pull-button {
-      width: 230px;
+      width: 100%;
     }
   }
 
   .toolbar-button {
     &.branch-toolbar-button {
-      width: 230px;
+      width: 100%;
```
Apply flexible width on the Branch and Push/Pull toolbar buttons.

jpedroso replied:
There’s another button just below named `.revert-progress` also with fixed width.

```
&.revert-progress {
    width: 230px;
}
```

I believe this button should also change to `width: 100%;` but haven’t yet found a way to get to a state where the button shows to confirm. Any ideas welcome.

tidy-dev replied:
This is seen when reverting a commit in a git LFS repo in which reverting a commit is very slow. Thus, this does need addressed. We can follow up with another PR since Git LFS is not a common usage and the likely hood that this would result in a negative experience is slim (worse case during loading the last toolbar button switches to 230px which could only be a clipping problem on a very small zoomed in screen). The `RevertProgress` toolbar button is rendered instead of the `PullPushFetch` so we can send the dimensions for the `PullPushFetch`  button down to it.

jpedroso committed:
Remove max width 65% for branch names

Branch names can expand to the available width in  `.branches-list-item`.

chronvas commented:
@jacortinas @sergiou87 Hello, can we hope to get that feature in? Any plans?

jpedroso commented:
@steveward @tidy-dev will this pr get any attention?

i ask because i found one layout problem with it -- the branch list grows beyond desired size when the current branch name is long enough, for which i can’t seem to find a quick fix --, but i will not put any more time into this if this just ends up stalled here.

tidy-dev commented:
@jpedroso Sorry for the delay and thank you for your patience. I cannot say a timeline, but I am interested in reviewing this. If you can find time to address the problem found that would be fantastic, (and if so, also updating this with the latest development would be awesome). Thank you for your time and patience on this.

jpedroso committed:
Use foldoutStyle to control min and max width

jpedroso committed:
New ToolboxDropdown.foldoutStyleOverrides property

Introduce the new property to avoid changing behavior of the existing `foldoutStyle` property.

Whereas `foldoutStyle` replaces default style entirely, `foldoutStyleOverrides` is used to add to or override the default style properties.

jpedroso commented:
@tidy-dev i fixed the remaining known issue, quoted below. also rebased with latest `development`. this is now ready for proper review.

>  the branch list grows beyond desired size when the current branch name is long enough, for which i can’t seem to find a quick fix 

Tried many things but couldn’t fix the issue entirely in stylesheets, so the way i fixed this was to pass additional style properties for the default styles that get returned in `ToolbarDropdown.getFoldoutStyle()`. I decided to preserve the existing functionality of `foldoutStyle` since it can be used to replace the default style entirely, and instead create a new property that adds to the default style, called `foldoutStyleOverrides`. Tried to convey this idea in the documentation best i could. 

Here’s the before and after when having a long branch name. 

#### Before (66570530d809b1cd67dd7674113889aee1a6afb3)
![2024-05-10-002917@2x](https://github.com/desktop/desktop/assets/3560/db83d507-1c85-4f5b-a1c7-22ee2f9250be)


#### After (9d7df2847a64aa14882f13086e164b55bbc01a80)
![2024-05-10-002916@2x](https://github.com/desktop/desktop/assets/3560/6c32e651-4704-4cea-98af-d5930d337570)

Jay-o-Way commented:
Helloooooo anybodyyyyy??

tidy-dev committed:
Merge branch 'development' into pr/18095

tidy-dev committed:
Make it so that the pull push fetch button isn't clipped on smallest screen and max zoom

tidy-dev reviewed (approved):
@jpedroso  Thank you for this PR and apologies for taking so long to review.

It works beautifully. 💖 

I kicked it around quite a bit and found an accessibility issue where on the min width screen and max zoom the fetch dropdown button was being clipped. Thus, I updated this to latest dev and added changes to make the branch dropdown and pull/push split the difference of the available screen width when it gets this limited size. Due to that I am going to get another engineer to review this before merging.

I also found that on max zoom and smallest windows, the branch list and pr list is clipped, but it is in production too.. just not as much so I don't think this PR should be held back because of it.

sergiou87 reviewed (approved):
Awesome work! 💖

tidy-dev commented:
Going to go ahead and merge. Tho we did discuss and decided we would like it behind a feature flag as it is such a prominent part of the app so we can have it on beta for a bit. I have got a working branch for that and will open a PR up shortly. It also addresses applying this logic to the revert commit toolbar button (which only applies to git LFS repos and only when reverting)

Jay-o-Way commented:
@tidy-dev glad this is finally picked up.
However I believe that a "feature flag" should be used for **new** features, not for a simple UI improvement.
